File: app/src/main/java/com/amazon/tv/firetv/leanbacklauncher/util/AppCategorizer.java
package com.amazon.tv.firetv.leanbacklauncher.util;


import android.content.pm.ActivityInfo;
import android.content.pm.ApplicationInfo;
import android.os.Build;

import com.amazon.tv.firetv.leanbacklauncher.apps.AppCategory;

public class AppCategorizer {

    public static AppCategory getAppCategory(String pkgName, ActivityInfo actInfo) {

        for (String s : VIDEO_FILTER) {
            if (pkgName.contains(s)) {
                return AppCategory.VIDEO;
            }
        }

        for (String s : MUSIC_FILTER) {
            if (pkgName.contains(s)) {
                return AppCategory.MUSIC;
            }
        }

        for (String s : GAMES_FILTER) {
            if (pkgName.contains(s)) {
                return AppCategory.GAME;
            }
        }

        if (actInfo != null) {
            if ((actInfo.applicationInfo.flags & ApplicationInfo.FLAG_IS_GAME) != 0 ||
                    (actInfo.applicationInfo.metaData != null && actInfo.applicationInfo.metaData.getBoolean("isGame", false))
            ) {
                return AppCategory.GAME;
            }
            if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                if (actInfo.applicationInfo.category == ApplicationInfo.CATEGORY_GAME) {
                    return AppCategory.GAME;
                } else if (actInfo.applicationInfo.category == ApplicationInfo.CATEGORY_VIDEO) {
                    return AppCategory.VIDEO;
                } else if (actInfo.applicationInfo.category == ApplicationInfo.CATEGORY_AUDIO) {
                    return AppCategory.MUSIC;
                }
            }
        }

        return AppCategory.OTHER;
    }

    public static final String[] VIDEO_FILTER = new String[]{
            "abema",
            "air.com.vudu",
            "android.rbtv",
            "appletv",
            "bbc.iplayer",
            "beeline.ott",
            "bloomberg",
            "cbs",
            "cn.tv",
            "com.canal.android",
            "com.google.android.tv",
            "com.lgi.orionandroid.tv",
            "com.nhl",
            "com.tvp.vod.tv",
            "com.vgtrk",
            "courville.nova",
            "ctcmediagroup.ctc",
            "dailymotion",
            "disney",
            "divantv",
            "dramafever",
            "dtv.tvx",
            "facebook",
            "firebat",
            "foxnews",
            "foxnow",
            "fujitv",
            "fxnow",
            "gyao",
            "haystack",
            "hbo",
            "hedwig",
            "hopwatch",
            "hulu",
            "imdb.tv",
            "iptvremote",
            "iqiyi",
            "khd.app",
            "kinopub",
            "kinotrend",
            "kodi",
            "lazycatsoft",
            "megafon.tv",
            "megogo",
            "moviesanywhere",
            "mts.mtstv",
            "mtv",
            "mubi",
            "netflix",
            "niklabs.pp",
            "ontv",
            "ott_play",
            "ottnavigator",
            "ottplay",
            "pl.redefine.ipla",
            "pl.tvn.player.tv",
            "planeta.tv",
            "raintv",
            "redbull.rbtv",
            "ru.beeline.tv",
            "ru.cn.tv",
            "ru.ctv",
            "ru.ivi",
            "ru.kinopoisk",
            "ru.ntv",
            "ru.rutube",
            "ru.start",
            "ru.tv1",
            "ru.yourok",
            "schabi",
            "serials",
            "showtime",
            "simon.kaelae",
            "skyking",
            "smarttubetv",
            "smotreshka",
            "spb.tv",
            "spbtv",
            "starz",
            "ted",
            "tnt_premier",
            "tricoloronline",
            "tubitv",
            "tv.a24h",
            "tv.kartina",
            "tv.mts",
            "tv.ntvplus",
            "tv.okko",
            "tv.pluto",
            "tv.vintera",
            "tver",
            "tviztv.tviz",
            "tvplayer",
            "twitch",
            "twittertv",
            "txcom.vplayer",
            "ua.youtv",
            "ufc.android",
            "unext",
            "video",
            "viki",
            "viu.tv",
            "viutv",
            "wifire",
            "youtube",
            "yuriev.ndr"
    };

    public static final String[] MUSIC_FILTER = new String[]{
            "audials",
            "audioplayer",
            "deezer",
            "fmplay",
            "ituner",
            "music",
            "pandora",
            "radio",
            "somafm",
            "spotify",
            "tidal",
            "tunein",
            "zaycev"
    };

    public static final String[] GAMES_FILTER = new String[]{
            "android.play.games",
            "signaltalk",
            "tetris"
    };

}

File: app/src/main/java/com/amazon/tv/firetv/leanbacklauncher/util/AppCategorizer.kt
package com.amazon.tv.firetv.leanbacklauncher.util

import android.content.pm.ActivityInfo
import android.content.pm.ApplicationInfo
import android.os.Build
import com.amazon.tv.firetv.leanbacklauncher.apps.AppCategory

object AppCategorizer {

    private val VIDEO_FILTER = arrayOf(
        "abema",
        "air.com.vudu",
        "android.rbtv",
        "anilabx3",
        "anixartd",
        "appletv",
        "bbc.iplayer",
        "beeline.ott",
        "bloomberg",
        "brouken.player",
        "catchplay",
        "cbs",
        "cn.tv",
        "com.canal.android.canal",
        "com.canal.android",
        "com.daserste.daserste",
        "com.google.android.tv",
        "com.lgi.orionandroid.tv",
        "com.nhl",
        "com.plexapp",
        "com.prosto.tv",
        "com.senchick.viewbox",
        "com.tvp.vodtv",
        "com.vgtrk",
        "com.viaplay.android",
        "com.zatoo.player",
        "courville.nova",
        "ctcmediagroup.ctc",
        "dailymotion",
        "de.swr.ard",
        "disney",
        "divantv",
        "domatv.app",
        "dramafever",
        "dtv.tvx",
        "dutafilm",
        "elevenpoland",
        "exoplayer",
        "facebook",
        "filimons.soap4allv",
        "filmbluray",
        "filmix",
        "filmsclub",
        "firebat",
        "foxnews",
        "foxnow",
        "fujitv",
        "fxnow",
        "gyao",
        "haystack",
        "hbo",
        "hedwig",
        "hopwatch",
        "hotstar",
        "hulu",
        "iflix",
        "imdb.tv",
        "iptvclient.android.idmnc",
        "iptvremote",
        "iptvsmarterstvbox",
        "iqiyi",
        "khd.app",
        "kinopub",
        "kinotrend",
        "kodi",
        "lazycatsoft",
        "lionsgateplay",
        "loklok",
        "maxstream",
        "megafon.tv",
        "megogo",
        "molatvbox",
        "movies",
        "mrmc",
        "mts.mtstv",
        "mtv",
        "mubi",
        "nathnetwork",
        "netflix",
        "niklabs.pp",
        "ontv",
        "ott_play",
        "ottclub",
        "ottnavigator",
        "ottplay",
        "pl.cda.tv",
        "pl.polsatgo",
        "pl.redefine.ipla",
        "pl.tvn.player.tv",
        "pl.tvp.tvp_sport",
        "planeta.tv",
        "player.clouddy.online",
        "player.online",
        "raintv",
        "redbull.rbtv",
        "ru.beeline.tv",
        "ru.cn.tv",
        "ru.ctv",
        "ru.ivi",
        "ru.kinopoisk",
        "ru.more.play",
        "ru.ntv",
        "ru.rutube",
        "ru.start",
        "ru.tv1",
        "ru.twicker.lampa",
        "ru.vokino",
        "ru.yourok",
        "schabi",
        "serials",
        "showtime",
        "simon.kaelae",
        "skyking",
        "smarttubetv",
        "smartup.tv",
        "smotreshka",
        "spb.tv",
        "spbtv",
        "starz",
        "ted",
        "tencent.qqlivei18n",
        "tiktok",
        "tnt_premier",
        "toggo",
        "tricoloronline",
        "tubitv",
        "tv.a24h",
        "tv.kartina",
        "tv.mola",
        "tv.mts",
        "tv.ntvplus",
        "tv.okko",
        "tv.pluto",
        "tv.vintera",
        "tv2x2",
        "tver",
        "tviztv.tviz",
        "tvplayer",
        "twitch",
        "twittertv",
        "txcom.vplayer",
        "ua.youtv",
        "ufc.android",
        "unext",
        "useetv",
        "uz.allplay.apptv",
        "uz.i_tv.player",
        "video",
        "videomanager.kids",
        "vidio.android",
        "viki",
        "visionplus",
        "viu.tv",
        "viutv",
        "vplay",
        "wifire",
        "wink",
        "youtube",
        "yuriev.ndr",
    )
    private val MUSIC_FILTER = arrayOf(
        "audials",
        "audioplayer",
        "deezer",
        "fmplay",
        "ituner",
        "music",
        "pandora",
        "radio",
        "somafm",
        "spotify",
        "tidal",
        "tunein",
        "zaycev",
    )
    private val GAMES_FILTER = arrayOf(
        "android.play.games",
        "signaltalk",
        "tetris",
        "com.google.stadia",
    )

    fun getAppCategory(pkgName: String?, actInfo: ActivityInfo?): AppCategory {
        pkgName?.let { pn ->
            for (s in VIDEO_FILTER) {
                if (pn.contains(s)) {
                    return AppCategory.VIDEO
                }
            }
            for (s in MUSIC_FILTER) {
                if (pn.contains(s)) {
                    return AppCategory.MUSIC
                }
            }
            for (s in GAMES_FILTER) {
                if (pn.contains(s)) {
                    return AppCategory.GAME
                }
            }
        }
        actInfo?.let { ai ->
            if (ai.applicationInfo.flags and ApplicationInfo.FLAG_IS_GAME != 0 ||
                ai.applicationInfo.metaData != null && ai.applicationInfo.metaData.getBoolean(
                    "isGame",
                    false
                )
            ) {
                return AppCategory.GAME
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                when (ai.applicationInfo.category) {
                    ApplicationInfo.CATEGORY_GAME -> {
                        return AppCategory.GAME
                    }
                    ApplicationInfo.CATEGORY_VIDEO -> {
                        return AppCategory.VIDEO
                    }
                    ApplicationInfo.CATEGORY_AUDIO -> {
                        return AppCategory.MUSIC
                    }
                    ApplicationInfo.CATEGORY_ACCESSIBILITY,
                    ApplicationInfo.CATEGORY_IMAGE,
                    ApplicationInfo.CATEGORY_MAPS,
                    ApplicationInfo.CATEGORY_NEWS,
                    ApplicationInfo.CATEGORY_PRODUCTIVITY,
                    ApplicationInfo.CATEGORY_SOCIAL,
                    ApplicationInfo.CATEGORY_UNDEFINED -> {
                        return AppCategory.OTHER
                    }
                }
            }
        }
        return AppCategory.OTHER
    }
}
